Add the release note for 1.32.0 (#6136)
line/armeria: site/src/pages/release-notes/1.32.0.mdx

---
date: 2025-03-04
---

## 🌟 New features

- **LoadBalancer**: You can now create a <type://LoadBalancer> using various load balancing strategies to
  select an element from a list of candidates. #5289 #5779
  ```java
  List<Endpoint> candidates = ...;
  LoadBalancer.ofRoundRobin(candidates);
  LoadBalancer.ofWeightedRoundRobin(candidates);
  LoadBalancer.ofSticky(candidates, contextHasher);
  LoadBalancer.ofWeightedRandom(candidates);
  LoadBalancer.ofRampingUp(candidates);
  ```
- **Filter Invalid IP address**: You can now reject invalid IP addresses using
  <type://ClientFactoryBuilder#ipAddressFilter(Predicate)>. #6106 #6111
  ```java
  ClientFactory
    .builder()
    .ipAddressFilter(ip -> { // 👈👈👈

>       if ("1.2.3.4".equals(ip.getAddress().getHostAddress())) {
>         return false;
>       }
>     return true;

Reviewer: Indent?
```suggestion
      if ("1.2.3.4".equals(ip.getAddress().getHostAddress())) {
        return false;
      }
      return true;
```

    }).build();
  ```
- **Graceful Shutdown**: You can now use <type://GracefulShutdown> to terminate unfinished requests
  when a server stops.  #5941
  ```java
  GracefulShutdown gracefulShutdown =
    GracefulShutdown
      .builder()
      .quietPeriod(Duration.ofSeconds(10))
      .timeout(Duration.ofSeconds(15))

>       .shutdownErrorFunction((ctx, req) -> {

Reviewer: The method name was changed.
```suggestion
      .toExceptionFunction((ctx, req) -> {
```
I forgot to update the PR description.

    .build();
  ```
- **Map Injection for an Annotated Service**: The annotated service now supports the `Map<String, Object>`
  parameter. #6058 #6072
  ```java
  @Get("/dogma/repos")
  public String repos(@Param Map<String, Object> repoOptions) {...}
  ```
- **Eureka InstanceInfo**: You can now retrieve a Eureka <type://InstanceInfo> from an <type://Endpoint>
  via <type://InstanceInfo#instanceInfo(Endpoint)>. #6056 #6069
  ```java
  EurekaEndpointGroup eurekaEndpointGroup = ...;
  eurekaEndpointGroup.endpoints().forEach(endpoint -> {
    InstanceInfo instanceInfo = InstanceInfo.instanceInfo(endpoint); // 👈👈👈
  ```
- **Client Preprocessor**: You can now use a client <type://Preprocessor> to customize certain properties. #6057
  ```java
  HttpPreprocessor preprocessor = (delegate, ctx, req) -> {
    ctx.setEndpointGroup(Endpoint.of("overriding-host"));
    return delegate.execute(ctx, req);
  };
  WebClient client =
    WebClient.builder()
             .preprocessor(preprocessor) // 👈👈👈
             .build();
  ```
- **Backoff Builder**: You can now easily create a <type://Backoff> using builder. #5483 #5488
  ```java

>   Backoff backoff2 =

Reviewer: ```suggestion
  Backoff backoff =
```

    Backoff.builderForExponential()
           .initialDelayMillis(100)
           .maxDelayMillis(10000)
           .multiplier(2.0)
           .build();
  ```

## 📈 Improvements

- Armeria now automatically registers the metrics of `PooledByteBufAllocator.DEFAULT`. #2633 #5916
- Server metrics now have a port tag for better monitoring and troubleshooting. #6089 #6116
- You can now see the remoteAddress in the error message when a <type://SessionProtocolNegotiationException>
  is raised. #6113
- You can now use the default <type://HealthCheckUpdateHandler> via <type://HealthCheckUpdateHandler#of()> #6002

## 🛠️ Bug fixes

- A cookie is correctly overwritten when a cookie with the same name, domain, and path is set. #5870 #6126
- <type://CorsPolicyBuilder#build()> is exposed publicly again. #6121
- Connections do not hold references on automatically aborted requests, resulting in less memory pressure. #6108
- <type://CircuitBreakerRuleBuilder#onTimeException()> and <Type://RetryRuleBuilder#onTimeoutException()>`
  now work when a connection or proxy establishment is timed out. #6107
- DNS refresh query is executed by the correct executor. #5891 #6092

> - You can now choose whether to use the `json_name` field option when transcoding HTTP/JSON to gRPC messages. #5890 #6082

Reviewer: Should we mention this in the breaking changes section as well?

Author: Forgot about that. Updated it. 😉

> - The Thrift module names are correctly set by replacing a dot followed by a digit with an underscore. #6075 #6076

Reviewer: Because it is not limited to Thrift. We also changed the automatic module name for Scala.
```suggestion
- `Automatic-Module-Name`s are correctly set by replacing a dot followed by a digit with an underscore. #6075 #6076
```
